lib/pbench/server/database/models/__init__.py

```diff
@@ -50,3 +51,44 @@ def process_result_value(self, value, dialect):
         if value is not None:
             value = value.replace(tzinfo=datetime.timezone.utc)
         return value
+
+
+class commonDBException(Exception):
+    """Generic base class for reporting errors ."""
+
+    def DuplicateValue(self):
+        """This method will be invoked if SQLAlchemy IntegrityError
+        throws UNIQUE constraint violation
+
+        ABSTRACT METHOD: override in subclass to perform operation.
+        """
+        raise NotImplementedError()
+
+    def NullKey(self):
+        """This method will be invoked if SQLAlchemy IntegrityError
+        throws NOT NULL constraint violation
+
+        ABSTRACT METHOD: override in subclass to perform operation.
+        """
+        raise NotImplementedError()
+
+    def _decode(self, exception: IntegrityError) -> Exception:
+        """Decode a SQLAlchemy IntegrityError to look for a recognizable UNIQUE
+        or NOT NULL constraint violation.
+
+        Return the original exception if it doesn't match.
+
+        Args:
+            exception : An IntegrityError to decode
+
+        Returns:
+            a more specific exception, or the original if decoding fails
+        """
+        # Postgres engine returns (code, message) but sqlite3 engine only
+        # returns (message); so always take the last element.
+        cause = exception.orig.args[-1]
+        if "UNIQUE constraint" in cause:
+            return self.DuplicateValue(self, cause)
+        elif "NOT NULL constraint" in cause:
+            return self.NullKey(self, cause)
+        return exception
```

Comment:
So, the situation here is that the knowledge of how to determine _which_ exception to raise should/needs to be encapsulated in this module, but the _identity_ (i.e., type) of the exceptions to be returned needs to be provided by the caller.  There are two obvious ways to accomplish that.

The first is for the caller to provide them as arguments to this function.  (They can be provided as instances, or as class objects, or as callables (e.g., factory functions).)  Under this approach, this `decode()` function doesn't need to be associated with any particular class -- it just needs to receive the exception instance to be decoded and the exception "ids" to return.

The second is to do as you've started to do here, to embed this function in the parent class of the exception instance to be decoded.  In this way, `self` will actually refer to a subclass instance, which means that this function (in the parent class) can invoke member functions in the subclass via `self` references.  But, I think it means that every instance of `self` has to know how to construct _both_ of the possibly-to-be-returned exception instances...which is slightly weird...and, then things start to unravel.

We have a number of class hierarchies, each descending from `Exception` via a separate custom base class (one for Audit, one for Server Settings, one for Templates, and one for Users), and they all derive a "duplicate" error and a "null-key" error from their respective base classes...which makes perfectly good sense, up until we start duplicating code.  I _think_ the ideal solution to this is a multiple inheritance approach (i.e., we have something which is, e.g., both an "Audit error" **and** an "SQL Constraint error").  But, it's not obvious to me how that factoid really helps us.

So, I'm inclined to say, for the immediate term, just use the first approach I outlined above:  make this `decode()` into a stand-alone utility function and have the caller just give it the two things that it should give back.  That will get us past the code duplication problem, and we can contemplate a more elegant solution at our leisure.

Reply:
I don't particularly like the idea of passing fully built exception object to be tossed conditionally; but it'd work, and it does avoid the need for consistent constructor semantics across the models. Then again, I'd prefer to have consistent DB exception format anyway, and I prefer my idea of passing class references to construct on the fly.

Reply:
I concur.  I think the right solution is to pass in factory methods which produce the required object when it's needed.  (It could just be a lambda expression on the caller's side.)  I don't much like the idea of passing in the class objects because it requires the decode function to assume that it knows what the c'tor signature is (and it requires all of the classes to have a common c'tor interface).